sdl2promela/builtins.py

```diff
@@ -24,29 +31,127 @@ def __translate_writeln(
 
 
 def __translate_length(
-    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Statement]
-) -> promelamodel.Statement:
-    """
-    Length assigns an integer scalar, so there is no need to pack it into assign_value call
-    """
+    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Expression]
+) -> promelamodel.Expression:
+    if len(parameters) != 1:
+        raise ValueError(
+            f"Length built-in function expected 1 arguments, but received {len(parameters)}."
+        )
+    if parameters[0] is None:
+        raise ValueError("Length queried entity is missing.")
+    if (
+        not isinstance(parameters[0], promelamodel.VariableReference)
+        and not isinstance(parameters[0], promelamodel.MemberAccess)
+        and not isinstance(parameters[0], promelamodel.ArrayAccess)
+    ):
+        raise ValueError("Invalid type of parameter for length")
+
+    return (
+        MemberAccessBuilder()
+        .withUtypeReference(parameters[0])
+        .withMember(VariableReferenceBuilder("length").build())
+        .build()
+    )
+
+
+def __translate_present(
+    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Expression]
+):
+    if len(parameters) != 1:
+        raise ValueError(
+            f"Present built-in function expected 1 argument, but received {len(parameters)}."
+        )
+    if parameters[0] is None:
+        raise ValueError("Present queried entity is missing.")
+    if (
+        not isinstance(parameters[0], promelamodel.VariableReference)
+        and not isinstance(parameters[0], promelamodel.MemberAccess)
+        and not isinstance(parameters[0], promelamodel.ArrayAccess)
+    ):
+        raise ValueError("Invalid type of parameter for present")
+
+    return (
+        MemberAccessBuilder()
+        .withUtypeReference(parameters[0])
+        .withMember(VariableReferenceBuilder("selection").build())
+        .build()
+    )
+
+
+def __translate_to_enum(
+    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Expression]
+):
+    if len(parameters) != 2:
+        raise ValueError(
+            f"to_enum built-in function expected 2 arguments, but received {len(parameters)}."
```

Comment:
Why 2? What happens with the second one? Should the subtracted value be 1 or e.g. min?

Reply:
Please add comment regarding to_enum scope and the fact that choice selectors are always indexed from 0. Please comment why -1.

Reply:
Please comment that this handled the two argument to_enum flavour, where first argument is the choice, and the second, ignored one, is the enum type.

Reply:
added